server/storage/storage_gc_test.go

```diff
@@ -0,0 +1,216 @@
+		"keySpace2",
+		"keySpace3",
+		"keySpace4",
+		"keySpace5",
+	}
+	safePoints := []uint64{
+		0,
+		1,
+		4396,
+		23333333333,
+		math.MaxUint64,
+	}
+	return spaceIDs, safePoints
+}
+
+func testServiceSafePoints() ([]string, []*endpoint.ServiceSafePoint) {
+	spaceIDs := []string{
+		"keySpace1",
+		"keySpace1",
+		"keySpace1",
+		"keySpace2",
+		"keySpace2",
+		"keySpace2",
+		"keySpace3",
+		"keySpace3",
+		"keySpace3",
+	}
+	expireAt := time.Now().Add(100 * time.Second).Unix()
+	serviceSafePoints := []*endpoint.ServiceSafePoint{
+		{ServiceID: "service1", ExpiredAt: expireAt, SafePoint: 1},
+		{ServiceID: "service2", ExpiredAt: expireAt, SafePoint: 2},
+		{ServiceID: "service3", ExpiredAt: expireAt, SafePoint: 3},
+		{ServiceID: "service1", ExpiredAt: expireAt, SafePoint: 1},
+		{ServiceID: "service2", ExpiredAt: expireAt, SafePoint: 2},
+		{ServiceID: "service3", ExpiredAt: expireAt, SafePoint: 3},
+		{ServiceID: "service1", ExpiredAt: expireAt, SafePoint: 1},
+		{ServiceID: "service2", ExpiredAt: expireAt, SafePoint: 2},
+		{ServiceID: "service3", ExpiredAt: expireAt, SafePoint: 3},
+	}
+	return spaceIDs, serviceSafePoints
+}
+
+func (s *testStorageGCSuite) TestSaveLoadServiceSafePoint(c *C) {
+	storage := NewStorageWithMemoryBackend()
+	testSpaceID, testSafePoints := testServiceSafePoints()
+	for i := range testSpaceID {
+		c.Assert(storage.SaveServiceSafePoint(testSpaceID[i], testSafePoints[i]), IsNil)
+	}
+	for i := range testSpaceID {
+		loadedSafePoint, err := storage.LoadServiceSafePoint(testSpaceID[i], testSafePoints[i].ServiceID)
+		c.Assert(err, IsNil)
+		c.Assert(loadedSafePoint, DeepEquals, testSafePoints[i])
+	}
+}
+
+func (s *testStorageGCSuite) TestLoadMinServiceSafePoint(c *C) {
+	storage := NewStorageWithMemoryBackend()
+	currentTime := time.Now()
+	expireAt1 := currentTime.Add(100 * time.Second).Unix()
+	expireAt2 := currentTime.Add(200 * time.Second).Unix()
+	expireAt3 := currentTime.Add(300 * time.Second).Unix()
+
+	serviceSafePoints := []*endpoint.ServiceSafePoint{
+		{ServiceID: "0", ExpiredAt: expireAt1, SafePoint: 100},
+		{ServiceID: "1", ExpiredAt: expireAt2, SafePoint: 200},
+		{ServiceID: "2", ExpiredAt: expireAt3, SafePoint: 300},
+	}
+
+	testKeySpace := "test"
+	for _, serviceSafePoint := range serviceSafePoints {
+		c.Assert(storage.SaveServiceSafePoint(testKeySpace, serviceSafePoint), IsNil)
+	}
+	minSafePoint, err := storage.LoadMinServiceSafePoint(testKeySpace, currentTime)
+	c.Assert(err, IsNil)
+	c.Assert(minSafePoint, DeepEquals, serviceSafePoints[0])
+
+	// this should remove safePoint with ServiceID 0 due to expiration
+	// and find the safePoint with ServiceID 1
+	minSafePoint2, err := storage.LoadMinServiceSafePoint(testKeySpace, currentTime.Add(150*time.Second))
+	c.Assert(err, IsNil)
+	c.Assert(minSafePoint2, DeepEquals, serviceSafePoints[1])
+
+	// verify that service safe point with ServiceID 0 has been removed
+	// removal might take some time
+	time.Sleep(time.Millisecond * 100)
```

Comment:
Could we remove it?

Reply:
The test will fail if we don't wait for some time for remove expired to finish
Since removal is now async from load min